INTEROP-8954: Prepare firewatch for Jira Cloud migration

amp-rh opened the pull request:
## Summary

Prepare firewatch for Jira Cloud migration (INTEROP-8954, INTEROP-8496):

- Switch user identification from username to accountId (`[~accountId:...]` mention format)
- Switch auth from Bearer token (`token_auth`) to Basic auth (`basic_auth`) — Bearer with API tokens returns 403 on Cloud
- Set REST API version to 3 (required for Cloud)
- Handle optional email (may be None due to Cloud privacy settings)
- Remove quotes from JQL issue key lookups (Cloud requires unquoted keys for standard fields)
- Add `--email` option to `jira-config-gen` CLI

### Changes by file

| File | Change |
|------|--------|
| `src/objects/jira_base.py` | Basic auth with email+token, API v3, unquoted JQL keys, backwards-compatible fallback to `token_auth` when email not configured |
| `src/escalation/jira_escalation.py` | accountId-based mentions and user comparison |
| `src/jira_config_gen/jira_config_gen.py` | Accept optional `email` parameter |
| `src/commands/jira_config_gen.py` | Add `--email` CLI option |
| `src/templates/jira.config.j2` | Include `email` field when provided |
| `tests/e2e/conftest.py` | Add `JIRA_EMAIL` env var fixture, update API endpoint to v3 |
| `tests/e2e/templates/jira_configs/jira.config.j2` | Include `email` field |
| `tests/unittests/conftest.py` | Mock config uses `email`+`token` (Basic auth), search mock handles `/search/jql` |
| `tests/unittests/.../fake_server_info.json` | Report `deploymentType: Cloud` |
| `tests/unittests/.../test_jira_escalation.py` | 7 tests for helper methods, accountId mention format |

### Pagination note

No code changes needed — python-jira 3.10.5 already detects `deploymentType: Cloud` and routes `search_issues()` to `enhanced_search_issues()`, which uses the `/search/jql` endpoint with `nextPageToken` pagination automatically.

## UAT follow-up

- [ ] `assign_issue` email lookup works in Cloud (`src/objects/jira_base.py`)
- [ ] `get_field("assignee")` string output displays correctly in issue tables
- [ ] Verify `JIRA_EMAIL` env var is set in CI secrets alongside `JIRA_TOKEN`

amp-rh committed:
Prepare Jira escalation for Cloud migration

Switch from username-based to accountId-based user identification
to support Jira Cloud migration:

- Use accountId for @mention syntax ([~accountId:...] format)
- Compare comment authors by accountId instead of email
- Add _get_user_account_id helper for safe accountId extraction
- Add _get_user_email helper with None handling for Cloud privacy
- Handle optional assignee_email (may be None in Cloud)

Includes 7 unit tests for the new helper methods.

amp-rh commented:
Gentle ping — this PR has been open since December 3 and is ready for review. It prepares Jira escalation for the Cloud migration by switching from username-based to accountId-based user identification (required for Jira Cloud). Could a maintainer take a look when available?

amp-rh committed:
Address Jira Cloud API migration gaps

- Switch auth from token_auth (Bearer) to basic_auth (email + API token)
  to avoid 403 errors on Jira Cloud. Falls back to token_auth when email
  is not configured, preserving backward compatibility.
- Set REST API version to 3 (required for Cloud).
- Remove quotes from JQL issue key lookups (Cloud requires unquoted keys
  for standard fields).
- Add --email option to jira-config-gen CLI and config templates.
- Update test fixtures: fake server info reports Cloud deployment type,
  mock HTTP handler matches /search/jql endpoint, e2e fixture uses
  /rest/api/3/issue.

Made-with: Cursor

sg-rh commented on src/commands/jira_config_gen.py:
Can we update the server address here

sg-rh replied:
I think there are multiple references to "https://issues.stage.redhat.com" (`JIRA_SERVER `/`server-url`)mainly in the documentation (readme, function params) that can be updated as well

amp-rh replied:
Updated all 10 references to `issues.stage.redhat.com` across 8 files (CLI help, docstrings, docs, test fixtures) to `https://redhat.stage.atlassian.net`.

sg-rh commented on src/objects/jira_base.py:
```suggestion
        For Jira Cloud (when email is set), uses the new search/jql API (enhanced_search_issues) when
        available to avoid the deprecated v2 search endpoint.
```

sg-rh replied:
maybe we can use the python-jira’s Cloud search library which has enhanced_search_issues (see [pycontribs/jira#2326](https://github.com/pycontribs/jira/pull/2326)), which uses /rest/api/3/search/jql and nextPageToken.

amp-rh replied:
python-jira 3.10.5 already handles this. `search_issues()` checks `self._is_cloud` and auto-routes to `enhanced_search_issues()` internally:

```python
if self._is_cloud:
    if startAt == 0:
        return self.enhanced_search_issues(...)
```

Calling `enhanced_search_issues` explicitly would be redundant and less portable. It's decorated with a `check_if_cloud` guard that returns `None` on Server/DC instances, so it would break backward compat. The current `search_issues()` call works transparently for both deployment types.

sg-rh replied:
okay great. Thanks for looking into this.

sg-rh commented on src/objects/jira_base.py:
```suggestion
        ""        if getattr(self, "email", None) and hasattr(self.connection, "enhanced_search_issues"):
            return self.connection.enhanced_search_issues(jql_query, maxResults=False)"
```

amp-rh committed:
Update Jira server URL references to Cloud staging instance

Replace all issues.stage.redhat.com references with
redhat.stage.atlassian.net across CLI help, docstrings,
docs, and test fixtures.

Made-with: Cursor

amp-rh committed:
Update src/objects/jira_base.py

Co-authored-by: sg-rh <sgill@redhat.com>

amp-rh committed:
Update src/objects/jira_base.py

Co-authored-by: sg-rh <sgill@redhat.com>

amp-rh committed:
Revert broken GitHub suggestions on search method

The accepted suggestions had malformed code (truncated triple-quote
with code on the same line). Also reverts the explicit
enhanced_search_issues call since python-jira 3.10.5 already
auto-routes search_issues() to enhanced_search_issues() on Cloud.

Made-with: Cursor

etirta commented on src/objects/jira_base.py:
```diff
@@ -128,7 +132,7 @@ def create_issue(
 
         if epic is not None:
-            epic_search = self.connection.search_issues(f'issue="{epic}"', maxResults=False)
+            epic_search = self.connection.search_issues(f"issue={epic}", maxResults=False)
```
Are yo sure this will work when `epic` has space or special characters? The original search query seems to indicate to enclosed within double-quotes (`"..."`) in order to allow space and/or special characters. Is this no longer the case with JIRA Cloud Query?

amp-rh replied:
Tested both forms live against `redhat.atlassian.net` (prod Cloud, `/rest/api/3/search/jql`):

| JQL | Result |
|-----|--------|
| `issue=INTEROP-1234` | Found |
| `issue="INTEROP-1234"` | Found |

Both work. Issue keys are structurally [`<PROJECT_KEY>-<NUMBER>`](https://docs.atlassian.com/software/jira/docs/api/10.0.0/com/atlassian/jira/issue/IssueKey.html) where the project key is `[A-Z0-9_]+` and the number is digits only. None of these characters are [JQL reserved characters](https://support.atlassian.com/jira-software-cloud/docs/what-is-advanced-search-in-jira-cloud/#Restricted-words-and-characters), so quoting is unnecessary. The [JQL fields doc](https://support.atlassian.com/jira-work-management/docs/jql-fields/) uses unquoted keys in its own examples (`workItemKey = ABC-123`). The original commit added the quotes without explanation, so dropping them to match the canonical form.

Added a unit test (`TestCreateIssueEpicSearch`) that asserts the JQL format.

etirta replied:
I see.. so the `epic` var. is guranteed to be JIRA Issue Key, and never the Issue Title? If so, then yes, we don't need to worry about quoting. Please confirm, thx.

amp-rh replied:
Confirmed. The `epic` parameter is always a Jira Issue Key (e.g. `PROJECT-123`), never a title. The value comes from the `jira_epic` field in firewatch rule config or the `FIREWATCH_DEFAULT_JIRA_EPIC` env var, both documented as issue keys. There's no code path that resolves titles to keys. If we want to add validation asserting the key format, we can track that in a separate issue.

amp-rh committed:
Add unit tests for Jira Cloud migration changes

Cover auth branching (basic_auth vs token_auth), API v3,
unquoted JQL epic search, and JiraConfig email rendering.

Made-with: Cursor

etirta reviewed (approved):
Approve.
